### PR TITLE
feat: rewrite sln in c# & upload extract to azure

### PR DESCRIPTION
**Pull request by YusufMavzer**

### CODE REVIEW

File: src/ExtractBundler.Console/Bundlers/BaseBundler.cs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+    private async Task GenerateZips()
+    {
+        //Stop eventHandlers
+        _extractDownloader.OnZipArchiveDownloaded -= ExtractDownloaderOnOnZipArchiveDownloaded;
+        _extractDownloader.OnZipArchiveDownloadFailed -= ExtractDownloaderOnOnZipArchiveDownloadFailed;
+
+        //Update MetaDataCenter
+        await _metadataClient.UpdateCswPublication(_identifier, DateTime.Now, _cancellationToken);
+        Console.WriteLine(_identifier.GetValue(ZipKey.MetadataUpdatedMessage));
```
Comment (jvandaal) [inline src/ExtractBundler.Console/Bundlers/BaseBundler.cs:78]: Shouldn't this be a log statement

Comment (YusufMavzer) [reply src/ExtractBundler.Console/Bundlers/BaseBundler.cs:78]: my bad, zal veranderen

File: src/ExtractBundler.Console/ExtractBundler.Console.csproj [hunk 2]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
+  <Import Project="..\..\packages\Be.Vlaanderen.Basisregisters.Build.Pipeline\Content\Be.Vlaanderen.Basisregisters.Build.Pipeline.Settings.App.props" />
+  <PropertyGroup>
+    <!-- Error NETSDK1152: https://docs.microsoft.com/en-us/dotnet/core/compatibility/sdk/6.0/duplicate-files-in-output -->
+    <ErrorOnDuplicatePublishOutputFiles>false</ErrorOnDuplicatePublishOutputFiles>
```
Comment (ArneD) [inline src/ExtractBundler.Console/ExtractBundler.Console.csproj:6]: Probably not needed here, no dependencies

File: src/ExtractBundler.Console/Bundlers/AddressBundler.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+namespace ExtractBundler.Console.Bundlers;
+
+using CloudStorageClients;
+using HttpClients;
+
+public class AddressBundler : BaseBundler
+{
+    public AddressBundler(ExtractDownloader extractDownloader, MetaDataCenterHttpClient metadataClient,
```
Comment (ArneD) [inline src/ExtractBundler.Console/Bundlers/AddressBundler.cs:9]: multiple lines

File: src/ExtractBundler.Console/Bundlers/BaseBundler.cs [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    private readonly ZipArchive _azureZipArchive;
+
+    private CancellationToken _cancellationToken;
+    readonly object _createEntryLock = new object();
+    private int _downloadedZipArchives = 0;
+    private bool _disposed = false;
+
+    public bool IsComplete { get; set; }
+    protected Identifier _identifier;
+    protected readonly List<string> _requiredZipArchives = new List<string>();
```
Comment (ArneD) [inline src/ExtractBundler.Console/Bundlers/BaseBundler.cs:35]: move above property

File: src/ExtractBundler.Console/Bundlers/FullBundler.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+namespace ExtractBundler.Console.Bundlers;
+
+using CloudStorageClients;
+using HttpClients;
+
+public class FullBundler : BaseBundler
+{
+    public FullBundler(ExtractDownloader extractDownloader, MetaDataCenterHttpClient metadataClient,
```
Comment (ArneD) [inline src/ExtractBundler.Console/Bundlers/FullBundler.cs:9]: multiple lines

File: src/ExtractBundler.Console/Bundlers/StreetNameBundler.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
+namespace ExtractBundler.Console.Bundlers;
+
+using CloudStorageClients;
+using HttpClients;
+
+public class StreetNameBundler : BaseBundler
+{
+    public StreetNameBundler(ExtractDownloader extractDownloader, MetaDataCenterHttpClient metadataClient,
+        S3Client s3Client, AzureBlobClient azureBlobClient) : base(extractDownloader, metadataClient, s3Client, azureBlobClient)
```
Comment (ArneD) [inline src/ExtractBundler.Console/Bundlers/StreetNameBundler.cs]: multiple lines

**Review by ArneD — CHANGES_REQUESTED** [review]

File: src/ExtractBundler/FileExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+using System.IO.Compression;
+using System.IO.Pipelines;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+public static class FileExtensions
+{
+    public static string EnsureDirectoryCreated(this string dirPath)
```
[the remark this answers is not in the document]
Comment (YusufMavzer) [reply src/ExtractBundler/FileExtensions.cs:14]: I'll remove it. Written during development for development

File: src/ExtractBundler.Console/CloudStorageClients/AzureBlobClient.cs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    using Azure.Storage.Blobs.Models;
+    using Azure.Storage.Blobs.Specialized;
+    using Infrastructure.Configurations;
+    using Microsoft.Extensions.Logging;
+    using Microsoft.Extensions.Options;
+
+    public class AzureBlobClient
+    {
+        private readonly AzureBlobOptions _options;
+        private const int chunkSizeInBytes = 4 * 1024 * 1024; // 4 MB chunk size
```
Comment (ArneD) [inline src/ExtractBundler.Console/CloudStorageClients/AzureBlobClient.cs:21]: Const naming Capital C

File: src/ExtractBundler.Console/CloudStorageClients/AzureBlobClient.cs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+            if (identifier == Identifier.AddressLinks)
+            {
+                return isTest ? $"31089/{blobName}" : $"10144/{blobName}";
+            }
+
+            return blobName;
+        }
+
+        public async Task UploadBlobInChunksAsync(byte[] content, Identifier identifier,
+            CancellationToken cancellationToken = default)
```
Comment (ArneD) [inline src/ExtractBundler.Console/CloudStorageClients/AzureBlobClient.cs:64]: multiple lines

File: src/ExtractBundler.Console/CloudStorageClients/AzureBlobClient.cs [hunk 10]
```diff
@@ -0,0 +1,137 @@
+﻿namespace ExtractBundler.Console.CloudStorageClients
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Azure;
+    using Azure.Identity;
+    using Azure.Storage;
```
Comment (ArneD) [inline src/ExtractBundler.Console/CloudStorageClients/AzureBlobClient.cs:10]: remove unused usings

File: src/ExtractBundler.Console/HttpClients/ExtractDownloadClient.cs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+using System.IO;
+using System.IO.Compression;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Infrastructure.Configurations;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+public sealed class ExtractDownloader
```
Comment (ArneD) [inline src/ExtractBundler.Console/HttpClients/ExtractDownloadClient.cs:14]: Filename <> Classname

File: src/ExtractBundler.Console/HttpClients/ExtractDownloadClient.cs [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+{
+    private readonly ILogger<ExtractDownloader>? _logger;
+    private readonly Dictionary<string, ApiEndPointOptions> _apiEndPointOptions;
+    private readonly Dictionary<string, ZipArchive> zipArchives = new();
```
Comment (ArneD) [inline src/ExtractBundler.Console/HttpClients/ExtractDownloadClient.cs:18]: `_zipArchives`

File: src/ExtractBundler.Console/HttpClients/ExtractDownloadClient.cs [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+            var archive = await GetZipArchiveWithDownloadProgressAsync(response!, fileName, cancellationToken);
+            zipArchives.Add(fileName, archive);
+            OnZipArchiveDownloaded?.Invoke((fileName, archive), EventArgs.Empty);
+            return;
+        }
+
+        OnZipArchiveDownloadFailed?.Invoke($"{fileName} : {url}", EventArgs.Empty);
+    }
+
+    private async Task<ZipArchive> GetZipArchiveWithDownloadProgressAsync(HttpResponseMessage response, string fileName,
```
Comment (ArneD) [inline src/ExtractBundler.Console/HttpClients/ExtractDownloadClient.cs:78]: multiple lines

File: src/ExtractBundler.Console/HttpClients/MetaDataCenterHttpClient.cs [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+using Infrastructure.Configurations;
+using Microsoft.Extensions.Configuration;
```
Comment (ArneD) [inline src/ExtractBundler.Console/HttpClients/MetaDataCenterHttpClient.cs:14]: unused

File: src/ExtractBundler.Console/Processors/ExtractBundleProcessor.cs [hunk 15]
```diff
@@ -0,0 +1,50 @@
+namespace ExtractBundler.Console.Processors;
+
+using System;
```
Comment (ArneD) [inline src/ExtractBundler.Console/Processors/ExtractBundleProcessor.cs:3]: unused

**Review by ArneD — CHANGES_REQUESTED** [review]

File: src/ExtractBundler.Console/ExtractBundler.Console.csproj (hunk 2 above)
Comment (ArneD) [reply src/ExtractBundler.Console/ExtractBundler.Console.csproj:5]: The line is still there :)

**Review by ArneD — APPROVED** [review]

**Pull request merged**